strider/server.py

```diff
@@ -46,78 +49,110 @@
+    qgraph = RedisGraph(f"{qid}:qgraph")
+    kgraph = RedisGraph(f"{qid}:kgraph")
+    results = RedisList(f"{qid}:results")
+    logs = RedisList(f"{qid}:log")
+
+    qgraph.expire(store_results_for)
+    kgraph.expire(store_results_for)
+    results.expire(store_results_for)
+    logs.expire(store_results_for)
+
+    return ReasonerResponse(
+        message=Message(
+            query_graph=qgraph.get(),
+            knowledge_graph=kgraph.get(),
+            results=list(results.get()),
+        ),
+        logs=list(logs.get()),
     )
 
 
-async def sync_answer(query: Dict, **kwargs):
-    """Answer biomedical question, synchronously."""
-    try:
-        queue = asyncio.PriorityQueue()
-        counter = itertools.count()
+async def process_query(qid):
+    # Set up workers
+    strider = StriderWorker(num_workers=2)
 
-        # setup strider
-        strider = StriderWorker(
-            queue,
-            num_workers=2,
-            counter=counter,
-        )
+    # Process
+    await strider.run(qid, wait=True)
 
-        await strider.run(query["message"]["query_graph"], wait=True)
-        return {
-            "query_graph": query["query_graph"],
-            "results": strider.results,
-            "knowledge_graph": strider.kgraph,
-        }
-    except NoAnswersError as err:
-        LOGGER.warning(str(err))
-        return query['message']
-    async with Database('results.db') as database:
-        return await _get_results(
-            query_id=query_id,
-            database=database,
-        )
+    # Pull results from redis
+    # Also starts timer for expiring results
+    return get_finished_query(qid)
 
 
-@APP.post('/aquery', response_model=str, tags=['query'])
+@APP.post('/aquery', tags=['query'])
 async def async_query(
-        query: Query,
-        support: bool = True,
-) -> str:
-    """Handle asynchronous query."""
-    query_id = await execute_query(
-        query.message.query_graph.dict(),
-        support=support,
-        wait=False,
+        background_tasks: BackgroundTasks,
+        query: Query = Body(..., example=EXAMPLE),
+) -> dict:
+    """Start query processing."""
+    # Generate Query ID
+    qid = str(uuid.uuid4())[:8]
+
+    # Save query graph to redis
+    qgraph = RedisGraph(f"{qid}:qgraph")
+    qgraph.set(query.dict()['message']['query_graph'])
+
+    # Start processing
+    background_tasks.add_task(process_query, qid)
+
+    # Return ID
+    return dict(id=qid)
+
+
+@APP.post('/query_result', response_model=ReasonerResponse)
+async def get_results(qid: str) -> ReasonerResponse:
+    print(get_finished_query(qid))
+    return get_finished_query(qid)
+
+
+@APP.post('/query', tags=['query'])
+async def sync_query(
+        query: Query = Body(..., example=EXAMPLE)
+) -> ReasonerResponse:
+    """Handle synchronous query."""
+    # Generate Query ID
+    qid = str(uuid.uuid4())[:8]
+
+    # Save query graph to redis
+    qgraph = RedisGraph(f"{qid}:qgraph")
+    qgraph.set(
```

Comment:
I just realized that we built an ORM, and that makes me feel icky. We access the query graph a lot; how much overhead is involved in querying redis every time? Since we basically never edit/delete things, can we cache the `get`s using e.g. lru?

Reply:
The reason there are so many ORMs is that they are constantly being spun off of other projects...

Reply:
I don't think we should worry about the overhead until we have tested and confirmed that it is an issue. Redis is in-memory and designed for very high throughput. We also have avenues for improvement such as caching using LRU. I don't think we need to implement those yet. 

As an aside, I think that we ended up with an ORM because this project benefits from having one. It's intuitive to structure web projects as an application layer interacting with a database.

Reply:
Okay.